### PR TITLE
disallowParenthesesAroundArrowParam: allow destructuring of param

### LINKED ISSUE #1672: disallowParenthesesAroundArrowParam should support ES2015(ES6) Destructuring Binding Patterns
screenshot

It seems that `disallowParenthesesAroundArrowParam` handles required parentheses for "[Destructuring Binding Patterns](http://www.ecma-international.org/ecma-262/6.0/#sec-destructuring-binding-patterns)" as needless parentheses, In JSCS 2.1.0.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/25712430-disallowparenthesesaroundarrowparam-should-support-es2015-es6-destructuring-binding-patterns?utm_campaign=plugin&utm_content=tracker%2F281640&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F281640&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by hzoo**
Fixes #1672

### CODE REVIEW

**Comment by markelog** [comment]
LGTM

File: test/specs/rules/disallow-parentheses-around-arrow-param.js [hunk 1]
```diff
@@ -29,4 +29,8 @@ describe('rules/disallow-parentheses-around-arrow-param', function() {
     it('should not report with parens around a multiple parameters with a default', function() {
         assert(checker.checkString('const a = (x = 1, y) => x * y;').isEmpty());
     });
+
+    it('should not report with use of destructuring #1672', function() {
```
Comment (mrjoelkemp) [inline test/specs/rules/disallow-parentheses-around-arrow-param.js]: Please add a test for destructuring usage that's not the first param: `(foo, {bar}) => ` is valid.

Comment (hzoo) [inline test/specs/rules/disallow-parentheses-around-arrow-param.js]: Ok - should fail first because of `node.params.length !== 1 ||`

**Comment by hzoo** [comment]
Refactored to used variables in the if statement

**Comment by mrjoelkemp** [comment]
Nice work! Merge away!

**Comment by hzoo** [comment]
Don't know how to do it yet - Although I'm assuming it's to rebase it and then add "Closes gh-1675"?

**Comment by mrjoelkemp** [comment]
Feel free to add the following to Maintain.md:

Some golden rules:

1. We **never** hit the Merge pull request button on github. This avoids merge commits which makes history muddy. We generate changelogs from the history, so the less noise – the better. 
1. We **never** change history on master. If you make a mistake and push up a silly commit to upstream's master, then you issue a revert commit, you **do not** remove the commit from the history and force push up.

My typical flow is (other maintainers likely have a more streamlined process):

1. `hub co the_pull_request_link` to check out the PR
2. Make sure that it's up to date with master: `git pull --rebase upstream master`
 - Assuming you have a remote `upstream` pointing to `git@github.com:jscs-dev/node-jscs.git`
3. Make sure the PR is still passing all tests: `npm test`
4. Fix up the commit message (according to the [suggested format](https://github.com/jscs-dev/node-jscs/blob/master/CONTRIBUTING.md#commit-message-format))
 - Make sure the issue number is listed if it's a bug fix: `Fixes #123`
 - Make sure the commit closes the PR number: `Closes gh-124`
 - Make sure any referenced issues are listed: `Ref #111`
 - If they have a single commit, then `git rebase -i HEAD` and choose to `reword` that commit according to the above rules.
 - If they have multiple commits, then `git rebase -i the_SHA_before_their_first_commit` and `fixup` all commits except the top most. `reword` the top-most commit according to the above rules.
5. Save the commit changes and exit the editor of that tab of the editor
6. The rebase should be carried out without a problem 
 - If you run into problems, `git rebase --abort` and ask another maintainer for help
7. `git checkout master && git pull upstream master && git merge -`
 - This should result in a *fast-forward* merge. If you have a recursive merge, then you did something wrong.
8. For sanity sake, feel free to run the test suite again post-merge: `npm test`
9. Once it's all passing, `git push upstream master` to push that commit to the master branch of the main jscs repo.

**Comment by hzoo** [comment]
Thanks for the in depth steps and explanation! Was mostly what I thought - add the `Closes` statement and fast forward merge it.

**Comment by mrjoelkemp** [comment]
Yes, please add to Maintain.md.

**Comment by SimenB** [comment]
Something about this (I think) broke jscs completely for us. `jscs@2.1.0` works fine, `jscs@2.1.1` explodes on any js file...

image

**Comment by hzoo** [comment]
@SimenB I can't reproduce with `jscs@2.1.1`

```
{
  "esnext": true,
  "disallowParenthesesAroundArrowParam": true
}
```

```
`import a from 'a';
```

**Comment by SimenB** [comment]
@hzoo This piece of code triggers the error: 

`window.onerror = () => {};`

Happens with zero parameters. Adding a parameter, with or without parentheses works fine

**Comment by hzoo** [comment]
Ok thanks will make a fix.

**Pull request closed without merging**